gargoyle-postgresql-connect/src/Gargoyle/PostgreSQL/Connect.hs

```diff
@@ -9,24 +9,32 @@ import Gargoyle (withGargoyle)
 import Gargoyle.PostgreSQL.Nix (postgresNix)
 import System.Directory (doesFileExist)
 
--- | Connects to a database using information at the given filepath
+
+-- Like 'withDb'' but turns the connection string into a connection 'Pool' for you.
+withDb :: String -> (Pool Connection -> IO a) -> IO a
+withDb dbPath f = withDb' dbPath (openDb >=> f)
+
+-- Convert a connection string into a connection 'Pool'.
+openDb :: ByteString -> IO (Pool Connection)
+openDb dbUri = createPool (connectPostgreSQL dbUri) close 1 5 20
+
+-- | Connects to a database using information at the given filepath.
 -- The given filepath can be either a folder (for a local db)
--- or a file with a database url
+-- or a file with a database connection string.
 --
--- withDb takes a String, which represents the path to a database, and a
+-- 'withDb'' takes a String, which represents the path to a database, and a
 -- function that returns database connection information as arguments in
 -- order to open and start the database. Otherwise, it will create the
 -- database for you if it doesn't exist.
-withDb :: String -> (Pool Connection -> IO a) -> IO a
-withDb dbPath a = do
+withDb' :: String -> (ByteString -> IO a) -> IO a
+withDb' dbPath f = do
   dbExists <- doesFileExist dbPath
   if dbExists
-    -- use the file contents as the uri for an existing server
-    then C8.readFile dbPath >>= openDb . head . C8.lines >>= a
+    -- use the file contents as the URI for an existing server
+    then C8.readFile dbPath >>= \b -> case C8.lines b of
+      [] -> error "DB connection string configuration file is empty"
```

Comment:
This is to avoid the use of `head` as in the previous version. We should probably also error-out if there are multiple lines because we silently ignore all but the first line. But that's a semantic change that might require version bump.

Reply:
Added a comment for this so it isn't lost.

Reply:
Do you think perhaps `withDb'` should return `IO (Either Text a)` so callers also have the option to handle invalid files some other way? Or at that point should `withDB'` stay the same way, and we instead have one function to parse the config / collect errors, and another to actually start the thing.

Reply:
Not a bad idea.